internal/generate/olm-catalog/csv.go

```diff
@@ -15,15 +15,436 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
 const (
-	OLMCatalogDir = scaffold.DeployDir + string(filepath.Separator) + "olm-catalog"
+	OLMCatalogChildDir = "olm-catalog"
+	// OLMCatalogDir is the default location for OLM catalog directory.
+	OLMCatalogDir  = scaffold.DeployDir + string(filepath.Separator) + OLMCatalogChildDir
+	csvYamlFileExt = ".clusterserviceversion.yaml"
+
+	// Input keys for CSV generator whose values are the filepaths for the respective input directories
+
+	// DeployDirKey is for the location of the operator manifests directory e.g "deploy/production"
+	DeployDirKey = "deploy"
+	// APIsDirKey is for the location of the API types directory e.g "pkg/apis"
+	APIsDirKey = "apis"
 )
 
+type csvGenerator struct {
+	gen.Config
+	// csvVersion is the CSV current version.
+	csvVersion string
+	// fromVersion is the CSV version from which to build a new CSV. A CSV
+	// manifest with this version should exist at:
+	// deploy/olm-catalog/{from_version}/operator-name.v{from_version}.{csvYamlFileExt}
+	fromVersion string
+	// existingCSVBundleDir is set if the generator needs to update from
+	// an existing CSV bundle directory
+	existingCSVBundleDir string
+	// csvOutputDir is the bundle directory filepath where the CSV will be generated
+	// This is set according to the generator's OutputDir
+	csvOutputDir string
+}
+
+func NewCSV(cfg gen.Config, csvVersion, fromVersion string) gen.Generator {
+	g := csvGenerator{
+		Config:      cfg,
+		csvVersion:  csvVersion,
+		fromVersion: fromVersion,
+	}
+	if g.Inputs == nil {
+		g.Inputs = map[string]string{}
+	}
+	// The olm-catalog directory location depends on where the deploy (operator
+	// manifests) directory is.
+	olmCatalogDir := OLMCatalogDir
+	if deployDir, ok := g.Inputs[DeployDirKey]; !ok || deployDir == "" {
+		g.Inputs[DeployDirKey] = scaffold.DeployDir
+	} else {
+		// Set to non-standard location.
+		olmCatalogDir = filepath.Join(g.Inputs[DeployDirKey], OLMCatalogChildDir)
+	}
+
+	parentDir := filepath.Join(olmCatalogDir, g.OperatorName)
+	if isBundleDirExist(parentDir, g.fromVersion) {
+		g.existingCSVBundleDir = filepath.Join(olmCatalogDir, g.OperatorName, g.fromVersion)
+	} else if isBundleDirExist(parentDir, g.csvVersion) {
+		g.existingCSVBundleDir = filepath.Join(olmCatalogDir, g.OperatorName, g.csvVersion)
+	}
+
+	if apisDir, ok := g.Inputs[APIsDirKey]; !ok || apisDir == "" {
+		g.Inputs[APIsDirKey] = scaffold.ApisDir
+	}
+	if g.OutputDir == "" {
+		g.OutputDir = scaffold.DeployDir
+	}
+
+	// Set the CSV bundle dir output path under the generator's OutputDir
+	g.csvOutputDir = filepath.Join(g.OutputDir, OLMCatalogChildDir, g.OperatorName, g.csvVersion)
+	return g
+}
+
+func isBundleDirExist(parentDir, version string) bool {
+	// Ensure full path is constructed.
+	if parentDir == "" || version == "" {
+		return false
+	}
+	bundleDir := filepath.Join(parentDir, version)
+	_, err := os.Stat(bundleDir)
+	if !os.IsNotExist(err) {
+		// TODO: return and handle this error
```

Comment:
> !os.IsNotExist(err) == os.IsExist(err) always

I don't think we can use these two checks interchangeably which is why I refactored this a bit. It depends on the kind of error you expect.

If a file already exist then `os.Stat(pathToExistingFile)` will not return an error of the type `os.IsExist(err)`. It will either be `err == nil`(if the file exists) or some other error(permissions, filesystem error etc).

So in `return err == nil || os.IsExist(err)`  `os.IsExist(err)` will always be false for `os.Stat()` so there's no reason to check it.

I can make this a bit more readable however with the following:
```Go
bundleDir := filepath.Join(parentDir, version)
_, err := os.Stat(bundleDir)
if os.IsNotExist(err) { // File does not exist
	return false
} else if err != nil { // some other more serious error
	// TODO: return and handle this error
	log.Fatalf("Failed to stat existing bundle directory %s: %v", bundleDir, err)
}
// File exists, with no error reported by os.Stat()
return true
```

Reply:
You are indeed [correct](https://gist.github.com/mastef/05f46d3ab2f5ed6a6787). This block should then be:
```go

func isBundleDirExist(parentDir, version string) (bool, error) {
	// Ensure full path is constructed.
	if parentDir == "" || version == "" {
		return false, nil
	}
	bundleDir := filepath.Join(parentDir, version)
	_, err := os.Stat(bundleDir)
	if err == nil || os.IsExist(err) {
		return true, nil
	}
	if !os.IsNotExist(err) {
		err = fmt.Errorf("error checking if bundle dir %q exists: %v", bundleDir, err)
	}
	return false, err
}
```

and be called in `Generate()` so the error can be returned.